src/execution_plan/optimizations/optimize_cartesian_product.h

```diff
@@ -0,0 +1,29 @@
+/*
+* Copyright 2018-2020 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#pragma once
+
+#include "../execution_plan.h"
+
+/* This optimization takes multiple branched cartesian product (with more than two branches),
+ * followed by filter(s) and try to apply the filter as soon possible by
+ * locating situations where a new Cartesian Product of just two streams can resolve the filter.
+ * can be filtered by the filter.
+ * For a filter F executing on a dual-branched cartesian product output, the runtime complexity is at most f=n^2.
+ * For a filter F' which execute on a dual-branched cartesian product output, where one of its branches is F,
+ * the overall time complexity is at most f'=f*n = n^3.
+ * In the general case, the runtime complaxity of filter that is executing over the output of a cartesian product
+ * which all of its children are nested cartesian product followed by a filter (as a result of this optimization)
+ * is at most n^x where x is the number of branchs of the original cartesian product.
+ * Consider MATCH (a), (b), (c) where a.x > b.x RETURN a, b, c
+ * Prior to this optimization a, b and c will be combined via a
+ * cartesian product O(n^3). Because we require a.v > b.v we can create a cartesian product between
```

Comment:
"will be combined via a cartesian product O(n^3)."
single line please.

Reply:
done